Add algorithms for getGamepads and events

nondebug opened the pull request:
Closes #114

The following tasks have been completed:

 * [ ] Modified Web platform tests (link to pull request)

Implementation commitment:

 * [ ] WebKit (https://bugs.webkit.org/show_bug.cgi?id=)
 * [ ] Chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=)
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/gamepad/pull/151.html" title="Last updated on Jul 8, 2021, 11:50 PM UTC (0abe48e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/gamepad/151/a1fdef2...0abe48e.html" title="Last updated on Jul 8, 2021, 11:50 PM UTC (0abe48e)">Diff</a>

nondebug committed:
Add algorithms for getGamepads and events

Fixes #114

nondebug committed:
use infra types, fix slots, etc

marcoscaceres commented on index.html:
```diff
@@ -170,6 +170,43 @@ <h2>
+          <td>A flag indicating that the {{Gamepad}} object has been exposed to script.
+        <tr>
+          <td><dfn>[[\axisMapping]]</dfn>
+          <td>`{}`
```
In the internal slots, we can switch to using infra types. That gives us nicer ways of working with things. 

```suggestion
          <td>An empty [=ordered map=]. 
```

nondebug replied:
Sounds good for these internal arrays. I've left the Navigator.gamepads list as an empty array [] since we need to expose that array to script for getGamepads. Is there some incantation for converting from infra types to Javascript types?

marcoscaceres commented on index.html:
```diff
@@ -170,6 +170,43 @@ <h2>
+        <tr>
+          <td><dfn>[[\axisMaximums]]</dfn>
+          <td>`null`
+          <td>An array containing the maximum logical value for each axis.
```
Probably want a list or [=ordered set=] here... 

```suggestion
          <td>A [=list=] containing the maximum logical value for each axis.
```

Same for the rest that use "array"...

marcoscaceres commented on index.html:
```diff
@@ -170,6 +170,43 @@ <h2>
+          <td>Mapping from unmapped axis index to an index in the {{Gamepad/axes}} array
+        <tr>
+          <td><dfn>[[\axisMinimums]]</dfn>
+          <td>`null`
```
Maybe better to just init to an empty list... then we don't need to check if it's null VS if it's empty.

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+      </ol>
+      To <dfn>update gamepad state</dfn> for |gamepad| with an <a>ordered array of logical axis input values</a> |axisValues:sequence&lt;unsigned long&gt;| and an <a>ordered array of logical button input values</a> |buttonValues:sequence&lt;unsigned long&gt;|, run the following steps:
+      <ol>
+        <li>Let |now:DOMHighResTimeStamp| be a {{DOMHighResTimeStamp}} representing the current time.
```
```suggestion
        <li>Let |now| be the [=current high resolution time=].
```

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+      To <dfn>update gamepad state</dfn> for |gamepad| with an <a>ordered array of logical axis input values</a> |axisValues:sequence&lt;unsigned long&gt;| and an <a>ordered array of logical button input values</a> |buttonValues:sequence&lt;unsigned long&gt;|, run the following steps:
+      <ol>
+        <li>Let |now:DOMHighResTimeStamp| be a {{DOMHighResTimeStamp}} representing the current time.
+        <li>Let |gamepadsStart:DOMHighResTimeStamp| be the |navigationStart:DOMHighResTimeStamp| attribute of the {{PerformanceTiming}} interface.
```
Hmm... this bit is not right for a couple of reasons:  

 1.  `PerformanceTiming` is just an interface (not an instance)
 2. `PerformanceTiming` was not initiated to anything. 
 3. `navigationStart` is "obsolete".

If we just need "when the global object was created", then:  
  
```suggestion
        <li>Let |gamepadsStart| be the [=time origin=] of the [=global object=].
```

Also, we probably don't need an instance of `DOMHighResTimeStamp` above.

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+      <ol>
+        <li>Let |now:DOMHighResTimeStamp| be a {{DOMHighResTimeStamp}} representing the current time.
+        <li>Let |gamepadsStart:DOMHighResTimeStamp| be the |navigationStart:DOMHighResTimeStamp| attribute of the {{PerformanceTiming}} interface.
+        <li>Set |gamepad|.{{Gamepad/timestamp}} to |now| − |gamepadsStart|.
```
```suggestion
        <li>Set |gamepad|.{{Gamepad/timestamp}} to |now| − |gamepadsStart|.
        <p class="warning">User agents SHOULD set a minimum resolution of |gamepad|’s {{Gampad/timestamp}} attribute to 5 microseconds, following [[HR-TIME]]'s [=clock resolution=] recommendation.
        </p>
```

nondebug replied:
I added the warning to the timestamp attribute definition

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+        <li>Set |gamepad|.{{Gamepad/timestamp}} to |now| − |gamepadsStart|.
+        <li>Let |axisCount:long| be the length of |axisValues|.
+        <li>Initialize |rawAxisIndex:long| to be 0.
+        <li>While |rawAxisIndex| < |axisCount|:
```
```suggestion
        <li>While |rawAxisIndex| is less than |axisCount|:
```
😇

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+        <li>Initialize |rawAxisIndex:long| to be 0.
+        <li>While |rawAxisIndex| < |axisCount|:
+          <ol>
+            <li>Let |mappedIndex:long| be the value for key |rawAxisIndex| in |gamepad|.{{[[axisMapping]]}}.
```
Because we are using an [=ordered map=], Infra let's us do: 

```suggestion
            <li>Let |mapping:ordered map| be |gamepad|.{{[[axisMapping]]}}.
            <li>Let |mappedIndex:long| be |mapping|["rawAxisIndex"].
```
And so on...

nondebug replied:
I've replaced the "value for key..." language with ordered map syntax but decided not to pull out |mapping|, maybe I should add it for readability? WDYT?

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+        <li>Let |buttonValues:sequence&lt;unsigned long&gt;| be the gamepad's <a>ordered array of logical button input values</a>.
+        <li>Queue a task (where?) to <a>update gamepad state</a> for |gamepad| with |axisValues| and |buttonValues|.
+      </ol>
+      To <dfn>update gamepad state</dfn> for |gamepad| with an <a>ordered array of logical axis input values</a> |axisValues:sequence&lt;unsigned long&gt;| and an <a>ordered array of logical button input values</a> |buttonValues:sequence&lt;unsigned long&gt;|, run the following steps:
```
I think we might want a generalized initializer algorithm for creating Gamepad instances. Check out how I did it in Geo:

https://www.w3.org/TR/geolocation/#constructing-a-geolocationposition

nondebug replied:
Added.

In another review it was pointed out that readonly attributes can't be updated after initialization (because they're read-only) so I've added internal slots for any attributes that may change after initialization and defined getter algorithms that refer to the internal slots.

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+          </ol>
+        <li>Let |buttonCount:long| be the length of |buttonValues|.
+        <li>Initialize |rawButtonIndex:long| to be 0.
+        <li>While |rawButtonIndex| < |buttonCount|:
```
```suggestion
        <li>While |rawButtonIndex| is less than |buttonCount|:
```

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+        <li>Initialize |rawButtonIndex:long| to be 0.
+        <li>While |rawButtonIndex| < |buttonCount|:
+          <ol>
+            <li>Let |mappedIndex:long| be the value for key |rawButtonIndex| in |gamepad|.{{[[buttonMapping]]}}.
```
Nit: Might be nice to pull these steps and the ones above into a two little algorithms. Pass in |gamepad| and it computes and stores the |normalizedValue|.

marcoscaceres commented on index.html:
```diff
@@ -258,6 +295,81 @@ <h2>
+              </ol>
+          </ol>
+      </ol>
+      <p>
+        The <a>user agent</a> MUST provide methods for retrieving <dfn data-lt="ordered array of axis inputs">ordered arrays of axis inputs</dfn> and <dfn data-lt="ordered array of button inputs">button inputs</dfn> for a gamepad such that there is a unique element in the axis array for each axis input and a unique element in the button array for each button input.
+        If the system enumerates axis and button inputs in a consistent order, then the methods SHOULD provide the inputs in that order.
+        Otherwise, the methods MAY use any consistent ordering.
+      </p>
+      <p>
+        The <a>user agent</a> MUST provide methods for retrieving <dfn data-lt="ordered array of minimum logical values">ordered arrays of minimum logical values</dfn> and <dfn data-lt="ordered array of maximum logical values">maximum logical values</dfn> for each gamepad input.
+        The ordering of these arrays MUST match the ordering of the <a>ordered array of axis inputs</a> and <a>ordered array of button inputs</a>.
+      </p>
+      <p>
+        The <a>user agent</a> MUST provide methods for retrieving <dfn data-lt="ordered array of logical axis input values">ordered arrays of logical axis input values</dfn> and <dfn data-lt="ordered array of logical button input values">button input values</dfn>.
+        These arrays MUST contain the most recent values received from the gamepad for each button and axis input.
+        The ordering of these arrays MUST match the ordering of the <a>ordered array of axis inputs</a> and <a>ordered array of button inputs</a>.
+      </p>
```
I'd suggest we delete these and just use Infra types. Also, it will avoid descriptors for things that happen algorithmically anyway.

marcoscaceres replied:
Just a note, only add a `data-lt=` to `dfn`s if you are creating an alias.

Like 
```
<dfn data-lt="fruit of the gods">bananas</dfn>  where the [=fruit of the gods=], and yet...
```

nondebug replied:
I've removed these definitions and rewritten the relevant part of the algorithm, please take a look

marcoscaceres commented on index.html:
```diff
@@ -358,68 +470,84 @@ <h2>
     <section data-dfn-for="Navigator">
       <h2>
-        Navigator Interface extension
+        Extensions to the {{Navigator}} interface
```
Just "code" (backticks) for headings is ok... the Navigator object will be hyperlinked in the IDL:
 
```suggestion
        Extensions to the `Navigator` interface
```

marcoscaceres commented on index.html:
```diff
@@ -358,68 +470,84 @@ <h2>
+      <p>
+        Instances of {{Navigator}} are created with the internal slots described in the following table:
+      </p>
+      <table class="simple" dfn-for=Navigator dfn-type=attribute>
```
```suggestion
      <table class="simple" dfn-for=Navigator>
```

marcoscaceres commented on index.html:
```diff
@@ -358,68 +470,84 @@ <h2>
+          <th>Initial value
+          <th>Description (non-normative)
+        <tr>
+          <td><dfn>[[\hasGamepadGesture]]</dfn>
```
These should be: 

```suggestion
          <td><dfn data-dfn-for="Navigator">[[\hasGamepadGesture]]</dfn>
```

And then... to link... a little ugly... I know: 

```
{{Navigator/[[hasGamepadGesture]]}}
```

marcoscaceres reviewed (commented):
Great stuff @nondebug! I didn't manage to get all the way through it yet, but here is the initial set of comments.

nondebug committed:
use infra lists for slots

nondebug committed:
use words instead of < and > for comparisons

marcoscaceres commented on index.html:
```diff
@@ -258,6 +278,80 @@ <h2>
+        <li>If |navigator|.{{Navigator/[[hasGamepadGesture]]}} is `false` and |gamepad| <a>contains a gamepad user gesture</a>:
+          <ol>
+            <li>Set |navigator|.{{Navigator/[[hasGamepadGesture]]}} to `true`.
+            <li>For each |connectedGamepad:Gamepad| of |navigator|.{{Navigator/[[gamepads]]}}:
```
Got motivated from our last tele conf and added nullable type support 😇

```suggestion
            <li>For each |connectedGamepad:Gamepad?| of |navigator|.{{Navigator/[[gamepads]]}}:
```

nondebug replied:
Thanks!

marcoscaceres commented on index.html:
```diff
@@ -258,6 +278,80 @@ <h2>
+                <li>If |connectedGamepad| is not `null`:
+                  <ol>
+                    <li>Set |connectedGamepad|.{{Gamepad/timestamp}} to |now|.
+                    <li>[[=Fire an event=]] named `"gamepadconnected"` at |window:Window| using {{GamepadEvent}} with its {{GamepadEvent/gamepad}} attribute initialized to |connectedGamepad|.
```
```suggestion
                    <li>[=Fire an event=] named `"gamepadconnected"` at |window:Window| using {{GamepadEvent}} with its {{GamepadEvent/gamepad}} attribute initialized to |connectedGamepad|.
```

nondebug replied:
Fixed

nondebug committed:
Tidy, syntax fixes

nondebug committed:
more changes

Replace <a>anchor tags</a> with [=definition links=]
Add internal slots and getters for readonly attributes
Add a Gamepad initializer algorithm
Add HR-TIME clock resolution warning
More Infra references and syntax
Fix gamepaddisconnected algorithm
Specify event targets
Add gh-contributors

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+                      <li>Set |connectedGamepad|.{{Gamepad/[[timestamp]]}} to
+                      |now|.
+                      </li>
+                      <li>If |gamepad|'s [=relevant global object=] is a
```
I think here we should be able to just queue on gamepad task source instead of going through the global one.

nondebug replied:
Done

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+                      |now|.
+                      </li>
+                      <li>If |gamepad|'s [=relevant global object=] is a
+                      {{Window}} object, [=queue a global task=] on |gamepad|'s
```
Here, we might want to also do the fully active check (as we are checking the {{Window}} is still there, per #149).... we can come back to this tho.

nondebug replied:
I've added a fully active check so we won't try to fire events for non-fully-active Documents.

I removed the Window check since I think we'll implicitly make that check when trying to get the associated (or responsible) Document. If the global object isn't a Window then it won't have a Document. Is that correct?

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+          following steps:
+        </p>
+        <ol>
+          <li>Let |axisValues:list| be a [=list=] of `unsigned long` values
```
```suggestion
          <li>Let |axisValues:list| be a [=list=] of {{unsigned long}} values
```

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+          the following steps:
+        </p>
+        <ol>
+          <li>Let |buttonValues:list| be a [=list=] of `unsigned long` values
```
```suggestion
          <li>Let |buttonValues:list| be a [=list=] of {{unsigned long}} values
```

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+          device represented by |gamepad|.
+          </li>
+          <li>Set |gamepad|.{{Gamepad/[[axisMinimums]]}} to a [=list=] of
+          `unsigned long` values with [=list/size=] equal to |inputCount|
```
```suggestion
          {{unsigned long}} values with [=list/size=] equal to |inputCount|
```

Plus a few other ones...

marcoscaceres commented on index.html:
```diff
@@ -195,69 +335,537 @@ <h2>
+          <li>Return |buttons|.
+          </li>
+        </ol>
+        <p>
+          User agents implementing this specification must provide a new DOM
+          event, named {{gamepadconnected}}. The corresponding event MUST be of
+          type {{GamepadEvent}} and, if [=allowed to use=] the `"gamepad"`
+          permission, MUST fire on the {{Window}} object. Registration for and
+          firing of the {{gamepadconnected}} event MUST follow the usual
+          behavior of DOM Events. [[DOM]]
+        </p>
+        <p>
+          A [=user agent=] MUST dispatch this event type to indicate the user
+          has connected a gamepad. If a gamepad was already connected when the
+          page was loaded, the {{gamepadconnected}} event SHOULD be dispatched
+          when the user presses a button or moves an axis.
+        </p>
+        <p>
+          A [=user agent=] MUST NOT dispatch this event type if the
+          [=environment settings object=] is a [=non-secure context=].
+        </p>
```
We don't need this stuff anymore 🎉... it's already above (including the secure context stuff... the API is not exposed in insecure contexts, so that can't ever happen). 

```suggestion
```

nondebug replied:
Oops, copy/paste error... removed

marcoscaceres commented on index.html:
```diff
@@ -275,43 +883,121 @@ <h2>
+        Instances of {{GamepadButton}} are created with the internal slots
+        described in the following table:
+      </p>
+      <table class="simple" dfn-for="GamepadButton">
```
```suggestion
      <table class="simple">
```

nondebug replied:
Removed

marcoscaceres commented on index.html:
```diff
@@ -722,47 +1502,116 @@ <h2>
+        steps:
+      </p>
+      <ol>
+        <li>If the [=current settings object=]'s [=environment settings object
+        / responsible document=] is not [=allowed to use=] the `"gamepad"`
+        permission, then abort these steps.
+        </li>
```
I don't think we need the check here? 

```suggestion
```

nondebug replied:
Removed.

marcoscaceres commented on index.html:
```diff
@@ -722,47 +1502,116 @@ <h2>
+        <li>Let |gamepad:Gamepad| be the {{Gamepad}} representing the
+        unavailable device.
+        </li>
+        <li>Set |gamepad|.{{Gamepad/[[connected]]}} to `false`.
```
I think this should be done in the steps of the task (right before the event is fired), no? Otherwise, it would update somewhat randomly.

nondebug replied:
Agreed. We should update the gamepad state synchronously with firing the event to make sure the state changes are handled sequentially and predictably.

I think this order makes the most sense for gamepadconnected:

1. Create gamepad with gamepad.connected initialized to true
2. Add gamepad to navigator.[[gamepads]]
3. Dispatch gamepadconnected

And for gamepaddisconnected:

1. Set gamepad.connected to false
2. Dispatch gamepaddisconnected
3. Remove gamepad from navigator.[[gamepads]]

Updating in this order ensures the gamepad.connected state matches the stated reflected by the gamepadconnected or gamepaddisconnected event, and also ensures the item in the getGamepads() array at index gamepad.index is the same instance as the gamepad instance in the event.

marcoscaceres commented on index.html:
```diff
@@ -722,47 +1502,116 @@ <h2>
-        object. Registration for and firing of the
-        <code>gamepaddisconnected</code> event MUST follow the usual behavior
-        of DOM Events. [[DOM]]
+        event, named {{gamepaddisconnected}}. The corresponding event MUST be
```
I think they "allowed to use" checks should only happen when the developer tries to use the API. 

We can maybe have a global check that if the permission to use the API gets disabled, then just bring down everything... but that's probably overkill.

nondebug replied:
I've removed the "allowed to use" check here but there's still a reference in the gamepadconnected algorithm.

marcoscaceres commented on index.html:
```diff
@@ -722,47 +1502,116 @@ <h2>
-        of DOM Events. [[DOM]]
+        event, named {{gamepaddisconnected}}. The corresponding event MUST be
+        of type {{GamepadEvent}} and, if [=allowed to use=] the `"gamepad"`
+        permission, MUST fire on the {{Window}} object. Registration for and
+        firing of the {{gamepaddisconnected}} event MUST follow the usual
+        behavior of DOM Events. [[DOM]]
```
Don't need this :) We are doing the right thing... 

```suggestion
        permission, MUST fire on the {{Window}} object.
```

nondebug replied:
Removed

marcoscaceres commented on index.html:
```diff
@@ -722,47 +1502,116 @@ <h2>
       <p>
-        A <a>user agent</a> MUST NOT dispatch this event type if the
-        <a>environment settings object</a> is a <a>non-secure context</a>.
+        A [=user agent=] MUST NOT dispatch this event type if the [=environment
+        settings object=] is a [=non-secure context=].
```
Don't need this... can't register events in insecure contexts, so they will never fire in these contexts... 

```suggestion
```

nondebug replied:
Removed

marcoscaceres reviewed (approved):
Wowser! This is so much better! Nice one @nondebug! 

I'd be inclined for us to take this as is (with the suggestions) and we should just build on it.

nondebug committed:
Update gamepad state on gamepad task source sequence

Changes the gamepadconnected/disconnected algorithms to update
gamepad state synchronously with firing the event.

Checks that the Document is fully active before firing events.

Removes some unnecessary language.

nondebug committed:
Merge branch 'gh-pages' into algorithms

marcoscaceres commented:
https://www.w3.org/TR/2021/WD-gamepad-20210709/ 🎉
